### PR TITLE
Adding 'MessageHasBeenDeleted' hook to plugins

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
As part of the MSTeams sync plugin I need to know when messages are deleted from the plugin. That makes a lot of sense because completes the whole plugin API around the posts.

#### Ticket Link
Not yet

#### Release Note
```release-note
Add new 'MessageHasBeenDeleted' hook to the plugin api
```

### CODE REVIEW

**Review by mickmister — APPROVED** [review]
:tada:

**Review by isacikgoz — APPROVED** [review]

**Comment by DHaussermann** [comment]
@jespino I'm not sure there is currently a way to validate this with functional testing. 
This would need to be added to the Demo Plugin. 

If this is urgent maybe there is a faster way to test it with a a tool like Postman or similar?
Long term, we would still want to open an issue to have this added to the demo plugin.

cc @mickmister

**Comment by mickmister** [comment]
> If this is urgent maybe there is a faster way to test it with a a tool like Postman or similar?

@DHaussermann This needs to be tested with a plugin that implements this hook. I've submitted https://github.com/mattermost/mattermost-plugin-demo/pull/159 for this. 

@jespino When testing this, I don't see this working correctly. I'm running this PR's code, and the demo plugin code linked above. For some reason the code is not entering the `RunMultiHook` block:

https://github.com/mattermost/mattermost-server/blob/0ca490ba59d62154e88e6e494b54239c48f94008/server/channels/app/post.go#L1320-L1325

File: server/plugin/hooks.go [hunk 1]
```diff
@@ -166,6 +167,13 @@ type Hooks interface {
 	// Minimum server version: 5.2
 	MessageHasBeenUpdated(c *Context, newPost, oldPost *model.Post)
 
+	// MessageHasBeenDeleted is invoked after the message has been deleted from the database.
+	// Note that this method will be called for posts deleted by plugins, including the plugin that
+	// deleted the post.
+	//
+	// Minimum server version: 7.9
```
Comment (hanzei) [inline server/plugin/hooks.go:174]: This needs to get updated

**Comment by jespino** [comment]
> > If this is urgent maybe there is a faster way to test it with a a tool like Postman or similar?
> 
> @DHaussermann This needs to be tested with a plugin that implements this hook. I've submitted [mattermost/mattermost-plugin-demo#159](https://github.com/mattermost/mattermost-plugin-demo/pull/159) for this.
> 
> @jespino When testing this, I don't see this working correctly. I'm running this PR's code, and the demo plugin code linked above. For some reason the code is not entering the `RunMultiHook` block:
> 
> https://github.com/mattermost/mattermost-server/blob/0ca490ba59d62154e88e6e494b54239c48f94008/server/channels/app/post.go#L1320-L1325

@mickmister can be this because the plugin is not compile to the branch mattemost-server version? The plugin itself need the new changes in the API client code, so if the plugin is not compiled using the branch here, it shouldn't find the hook.

**Comment by DHaussermann** [comment]
@jespino and @mickmister for what it's worth - U also don'e see the message deleted hook working as expected when I have https://github.com/mattermost/mattermost-plugin-demo/pull/159 running on the server. 

I got latest `master` before compiling the demo plugin PR.

**Comment by jespino** [comment]
@DHaussermann that is the thing, latest master is not enough, you need this branch compiled for the server, and the plugin compiled against this branch code for the plugin library.

**Comment by DHaussermann** [comment]
Thanks @jespino I'll try that 👍

**Comment by mickmister** [comment]
@DHaussermann @jespino I've created https://github.com/mattermost/mattermost-plugin-starter-template/pull/182 to test this. I've updated the plugin to use `github.com/mattermost/mattermost-server/server/public`, and have used a `replace` statement in the `go.mod` file to use my local branch of `mattermost-server`, which is currently the latest commit on the branch for this PR. However, I'm still not seeing the post deletion behavior.

**Comment by mickmister** [comment]
@jespino Maybe you could try creating a minimal reproducible example for testing?

File: server/public/plugin/hooks.go [hunk 2]
```diff
@@ -166,6 +167,13 @@ type Hooks interface {
 	// Minimum server version: 5.2
 	MessageHasBeenUpdated(c *Context, newPost, oldPost *model.Post)
 
+	// MessageHasBeenDeleted is invoked after the message has been deleted from the database.
+	// Note that this method will be called for posts deleted by plugins, including the plugin that
+	// deleted the post.
+	//
+	// Minimum server version: 7.10
```
Comment (hanzei) [inline server/public/plugin/hooks.go:174]:
```suggestion
	// Minimum server version: 8.0
```

**Comment by jespino** [comment]
@DHaussermann I updated this to the latest version, I going to pass you a very simple sample plugin to test it

**Comment by hanzei** [comment]
@jespino Would you minding adding a simple example to the demo plugin?

**Comment by jespino** [comment]
@hanzei on it :)

**Comment by jespino** [comment]
@hanzei actually, there was already a PR for that, so I did it but I remove it in favor of the previous existing one.

File: server/public/plugin/hooks.go (hunk 2 above)
Comment (jespino) [reply server/public/plugin/hooks.go]: Updated to 9.1

**Comment by hanzei** [comment]
Sorry @jespino, I missed that @mickmister already created a PR.

@mickmister Your PR https://github.com/mattermost/mattermost-plugin-starter-template/pull/182 does heavily conflict with `master`. Would you remind resolving the conflicts so we can move forward with @jespino's PR?

**Comment by mickmister** [comment]
@hanzei The mentioned PR is in the demo plugin https://github.com/mattermost/mattermost-plugin-demo/pull/159, whereas the one linked above is in the starter template. I think the starter template one was just made to see if it compiled? Not sure. I've closed that one now since there's no reason for that code to be in the starter template.

I think the demo plugin PR needs to update its server dependency for this PR. That may be necessary when using new hooks. Are you able to update the dependency on that PR?

**Comment by mickmister** [comment]
@hanzei Alternatively, we can favor https://github.com/mattermost/mattermost-plugin-demo/pull/165 over that PR so @jespino can cover any changes needed there

**Comment by hanzei** [comment]
@mickmister I confused the multiple PRs. You are right; https://github.com/mattermost/mattermost-plugin-demo/pull/159 is ready for testing.

> I think the demo plugin PR needs to update its server dependency for this PR. That may be necessary when using new hooks. Are you able to update the dependency on that PR?

I've used a `replace` directive in the past.

If @DHaussermann is fine with that, I'm happy to test https://github.com/mattermost/mattermost-plugin-demo/pull/159 and this PR together, so we can move forward.

**Comment by hanzei** [comment]
Tested with https://github.com/mattermost/mattermost-plugin-demo/pull/165:
- Deletion message gets posted for deleted posts in public channels, private channels and DMs
- The posted message contains a working link to the channel and the user

**Comment by hanzei** [comment]
@jespino One thought before we merge this PR: Would we add an `actor *model.User` to the hook's signature that contains the user that deletes the posts? That might be helpful for reporting tools.

**Comment by hanzei** [comment]
Or does it make sense to (in a separate ticket) add `UserId` to `plugin.Context`?

**Pull request merged**